src/daft-plan/src/ops/project.rs

```diff
@@ -22,20 +25,275 @@ impl Project {
         projection: Vec<Expr>,
         resource_request: ResourceRequest,
     ) -> Result<Self> {
-        let upstream_schema = input.schema();
+        // Factor the projection and see if there are any substitutions to factor out.
+        let (factored_input, factored_projection) =
+            Self::try_factor_subexpressions(input, projection, &resource_request)?;
+
+        let upstream_schema = factored_input.schema();
         let projected_schema = {
-            let fields = projection
+            let fields = factored_projection
                 .iter()
                 .map(|e| e.to_field(&upstream_schema))
                 .collect::<common_error::DaftResult<Vec<_>>>()
                 .context(CreationSnafu)?;
             Schema::new(fields).context(CreationSnafu)?.into()
         };
         Ok(Self {
-            input,
-            projection,
+            input: factored_input,
+            projection: factored_projection,
             resource_request,
             projected_schema,
         })
     }
+
+    fn try_factor_subexpressions(
+        input: Arc<LogicalPlan>,
+        projection: Vec<Expr>,
+        resource_request: &ResourceRequest,
+    ) -> Result<(Arc<LogicalPlan>, Vec<Expr>)> {
+        // Given construction parameters for a projection,
+        // see if we can factor out common subexpressions.
+        // Returns a new set of projection parameters
+        // (a maybe new input node, and a maybe new list of projection expressions).
+        let upstream_schema = input.schema();
+        let (projection, substitutions) = Self::factor_expressions(&projection, &upstream_schema);
+
+        // If there are substitutions to factor out,
+        // create a child projection node to do the factoring.
+        let input = if substitutions.is_empty() {
+            input
+        } else {
+            let child_projection = projection
+                .iter()
+                .flat_map(optimization::get_required_columns)
+                .map(|colname| {
+                    substitutions.get(&colname).map_or_else(
+                        || (colname.clone(), Expr::Column(colname.clone().into())),
+                        |expr| {
+                            (
+                                colname.clone(),
+                                Expr::Alias(expr.clone().into(), colname.clone().into()),
+                            )
+                        },
+                    )
+                })
+                .collect::<IndexMap<_, _>>()
+                .into_values()
+                .collect::<Vec<_>>();
```

Comment:
Nit: We can collect the required columns into an `IndexSet` and then go directly to the `Expr` vec, which elides tuple creation and the like.
```suggestion
            let child_projection = projection
                .iter()
                .flat_map(optimization::get_required_columns)
                .collect::<IndexSet<_>>()
                .into_iter()
                .map(|colname| {
                    substitutions.get(&colname).map_or_else(
                        || Expr::Column(colname.clone().into()),
                        |expr| Expr::Alias(expr.clone().into(), colname.clone().into()),
                    )
                })
                .collect::<Vec<_>>();
```